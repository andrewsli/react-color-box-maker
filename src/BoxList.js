import React, { Component } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import uuid from 'uuid/v4';


class BoxList extends Component {
  constructor(props) {
    super(props);
    this.state = { boxes: [] };
    this.addBox = this.addBox.bind(this);
    this.removeBox = this.removeBox.bind(this);
  }

  //{width, height, color,} => this.setState(every box plus the one entered)
  addBox(box){
    let newBox = {...box, id: uuid(), key: uuid()};
    this.setState(state =>({
      boxes: [...state.boxes, newBox]
    }))
  }

  //box ID => this.setstate(keep every box except the one with matching ID)
  removeBox(boxId){
    this.setState(st => ({
      boxes: st.boxes.filter(box => box.id !== boxId)
    }));
  }
  
  render() {
    let boxes =
    <div>
      {this.state.boxes.map(
        box => <Box
        height={box.height}
        width={box.width}
        backgroundColor={box.backgroundColor}
        id={box.id}
        key={box.id}
        removeThisBox={this.removeBox}/>
      )}
    </div>
    
    return(
      <div>
        {boxes}
        <br/>
        <NewBoxForm addNewBox={this.addBox}/>
      </div>
    )
  }
}

export default BoxList;