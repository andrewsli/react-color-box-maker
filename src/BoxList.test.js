import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import BoxList from "./BoxList";


it("renders without crashing", function () {
    shallow(<BoxList />);
});


it("matches snapshot", function () {
    let wrapper = shallow(<BoxList />);
    let serialized = toJson(wrapper);
    expect(serialized).toMatchSnapshot();
});

it("adds boxes", function () {
    const boxes = [{ height: "100px", width: "100px", backgroundColor: "green"}];
    const wrapper = mount(<BoxList />);

    //test that we are starting with inital state
    wrapper.setState({ boxes });
    expect(wrapper.state().boxes).toEqual([
        { height: "100px", width: "100px", backgroundColor: "green"}
    ]);

    //test state after adding box
    wrapper.instance().addBox({ height: "50px", width: "50px", backgroundColor: "blue", key:"1"});
    expect(wrapper.state().boxes).toEqual([
        { height: "100px", width: "100px", backgroundColor: "green" },
        { height: "50px", width: "50px", backgroundColor: "blue", id: expect.any(String), key: expect.any(String) }
    ]);
})

it("finds the form and successfully creates boxes", function () {
    const wrapper = mount(<BoxList />);

    //get all our input fields
    const heightInput = wrapper.find("#height");
    const widthInput = wrapper.find("#width");
    const bgColorInput = wrapper.find("#backgroundColor");
    const form = wrapper.find("form");

    //change all our fields and submit the form
    heightInput.instance().value = "100";
    heightInput.simulate("change");

    widthInput.instance().value = "100";
    widthInput.simulate("change");

    bgColorInput.instance().value = "green";
    bgColorInput.simulate("change");

    form.simulate("submit");

    expect(wrapper.find(".box").last().html()).toContain('height: 100px; width: 100px; background-color: green;');
})