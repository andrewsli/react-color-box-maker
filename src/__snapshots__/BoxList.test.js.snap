// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div />
  <br />
  <NewBoxForm
    addNewBox={[Function]}
  />
</div>
`;
